### PR TITLE
Implemented add_event and run in NetworkSimulator

### PR DESCRIPTION
**Pull request by hatooku**

### CODE REVIEW

File: networksimulator.py [hunk 1]
```diff
@@ -52,6 +52,14 @@ def cur_time(self, time):
         self._cur_time = time
 
     @property
+    def num_active_flows(self):
+        return self._num_active_flows
+    
+    @num_active_flows.setter
+    def num_active_flows(self, num):
```
Comment (Jdonerme) [inline networksimulator.py]: Should this be a global setter?

File: networksimulator.py [hunk 2]
```diff
@@ -189,22 +198,29 @@ def run(self, duration=sys.float_info.max):
         """Runs the simulation for the given duration.
         
         Args:
-            duration (float): the duration of the simulation in milliseconds.
+            duration (float): the duration of the simulation in seconds.
                 By default, the simulation runs until termination.
 
         """
-        pass
+        while self.pq and self.num_active_flows > 0 and self.cur_time < duration:
```
Comment (Jdonerme) [inline networksimulator.py]: Shouldn't this be len(self.pq)?

File: networksimulator.py [hunk 3]
```diff
@@ -116,6 +124,10 @@ def add_flow(self, flow_id, src, dest, data_amount):
 
         flow = Flow(self, flow_id, src, dest, num_bits)
         self.flows[flow_id] = flow
+        self._num_active_flows += 1
+
+    def decrement_active_flows(self):
```
Comment (malavika1) [inline networksimulator.py]: why does this exist?

File: networksimulator.py (hunk 2 above)
Comment (malavika1) [inline networksimulator.py]: i don't really remember how conditions work, but i'm not convinced that this is right

File: networksimulator.py (hunk 1 above)
Comment (hatooku) [inline networksimulator.py]: Fixed

File: networksimulator.py (hunk 3 above)
Comment (hatooku) [inline networksimulator.py]: Flows need to decrement num_active_flows when they're done sending packets

File: networksimulator.py (hunk 2 above)
Comment (hatooku) [inline networksimulator.py]: len(self.pq) and self.pq act the same way in this conditional

**Pull request merged**